### ISSUE TITLE
Geoserver Extension

### ISSUE THREAD

**Issue by kmarszalek** [cyfronet-fid/sat4envi#340]
Something about GeoTools custom data store: http://docs.geotools.org/latest/userguide/tutorial/datastore/index.html

**Comment by jswk** [comment]
Security considerations: #108.

**Comment by jswk** [comment]
The [WMS 1.1.1 specification](https://portal.opengeospatial.org/files/?artifact_id=1081&version=1&format=pdf) includes custom dimensions, apart from standard `time` and `elevation`. We may use them to pass the scene identification. (Although they may be used alongside `time`, this is for consideration.)

The parameters are populated in [WMSRequests](https://github.com/geoserver/geoserver/blob/22d4c332a329c52a8cdf1e4be74c4946c7e0e860/src/wms/src/main/java/org/geoserver/wms/WMSRequests.java#L371)

**Comment by kmarszalek** [comment]
@jswk : https://docs.cyfronet.pl/display/FID/GeoServer+z+imagemosaic+i+S3

**Comment by jswk** [comment]
1. Dockerize the version of GeoServer with required fix.
2. Test imagemosaic layer with geostationary data: ~1m granules. (1 scene and geotiff per timestamp.)
3. Test imagemosaic layer with Sentinel data. First, fewer granules, but there will be a number of scenes per timestamp.

ad. 1. Use the forked version of GeoTools when building the docker image.
Publish to hub as `1.1.1-GS2.16.2`.

ad. 2. Some things to remember:
* enable postgis extension in db (use a proper docker image)
* add fields `geometry` and `location` to table `scene`
* calculate the footprint of the granules (might be enough to do it once for a product)
* see if there isn't a loop over all the granules when creating a layer, or accessing it after startup (?)

ad. 3. Some things to remember:
* again, footprints. but this time they have to be populated per granule
* see how the transparency is handled, so that the granules are stitched correctly
* bigger files, how is it performing?

**Comment by jswk** [comment]
ad. 1.: done, but with version [`1.2.0-GS2.16.2`](https://registry.hub.docker.com/layers/fiddev/geoserver/1.2.0-GS2.16.2/images/sha256-7c5b742bec24bfbf25ea8367f98428e4bd1e7650b89cbaca414158c72b36b92b?context=explore)

**Comment by jswk** [comment]
ad. 2.:
See branch [im-seed](https://github.com/cyfronet-fid/sat4envi/tree/im-seed).
So far, I only tested using an extra table for granules, which doesn't integrate with existing data.
The footprints were generated by IM itself, however, in practice we will get them from metadata.
The test dataset comprised of 350k granules, each with different timestamp, same footprint, which is equivalent to 10 years of scenes every 15 minutes.
There was no performance issue from GS or db. Only the time extent presentation in getCapabilities has to be set appropriately (not to list every available granule).

ad. 3.:
I tested against Sentinel-2 and Sentinel-1 GRDH data.
The geotiffs must have proper nodata property set, so that it is possible to compose them (-a_nodata 0 switch in gdal_transform).
S2 mosaic had 29 granules (same timestamp) - each around 300MB, GRDH is heavier: 1.2GB, but only three granules were tested.
Using setup with local GS and external CEPH storage on wired, the performance was fine.

Next: I'm going to address db structure to fit both IM and scene/metadata files.
